File: 7 kyu/Sum a list but ignore any duplicates.py
'''
https://www.codewars.com/kata/5993fb6c4f5d9f770c0000f2

Please write a function that sums a list, but ignores any duplicate items in the list.

For instance, for the list [3, 4, 3, 6] , the function should return 10.
'''

# my solution
def last_index_of(list, value):
    return len(list) - list[::-1].index(value) - 1

def sum_no_duplicates(l):
    sum = 0
    for idx, el in enumerate(l):
        if l.index(el) == last_index_of(l, el):
            sum += el
    return sum

#! alternative-solution-1
from collections import Counter

def sum_no_duplicates(nums):
    return sum(k for k, v in Counter(nums).items() if v == 1)

#! alternative-solution-2
def sum_no_duplicates(l):
    return sum(n for n in set(l) if l.count(n) == 1)

File: Java/7kyu/Sum a list but ignore any duplicates.java
/*
https://www.codewars.com/kata/5993fb6c4f5d9f770c0000f2

Please write a function that sums a list, but ignores any duplicate items in the list.

For instance, for the list [3, 4, 3, 6] , the function should return 10.
*/

// my solution
import java.util.*;


public class Solution{
    public static int sumNoDuplicates(int[] arr){
        int sum = 0;
        
        for (int i : arr) {
            if (Arrays.stream(arr).filter(c -> c == i).count() == 1)
            sum += i;
        }
        
        return sum;
    }
}

//! alternative-solution
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Solution{
    public static int sumNoDuplicates(int[] arr){
        Set<Integer> setOfDuplicates = new HashSet<>();
        Set<Integer> setOfUniques = new HashSet<>();
            
        for (int i : arr){
                if (!setOfUniques.add(i)){
                    setOfDuplicates.add(i);
                }
        }
            
        return Arrays.stream(arr).filter(i -> !setOfDuplicates.contains(i)).sum();
    }
}